msal/src/main/java/com/microsoft/identity/client/Account.java

```diff
@@ -22,73 +22,102 @@
+import android.support.annotation.Nullable;
+
+import com.microsoft.identity.common.internal.logging.Logger;
+import com.microsoft.identity.common.internal.providers.microsoft.MicrosoftIdToken;
+import com.nimbusds.jwt.JWT;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.JWTParser;
+
+import java.text.ParseException;
+import java.util.HashMap;
+import java.util.Map;
+
+public class Account implements IAccount {
+
+    private static final String TAG = Account.class.getSimpleName();
+
+    private final String mRawIdToken;
+    private String mHomeOid;
+    private String mHomeTenantId;
     private String mEnvironment;
 
-    Account() {
-        // Empty constructor
+    public Account(@Nullable final String rawIdToken) {
+        mRawIdToken = rawIdToken;
     }
 
-    /**
-     * Sets the account id.
-     *
-     * @param accountId The IAccountIdentifier to set.
-     */
-    void setAccountIdentifier(final IAccountIdentifier accountId) {
-        mAccountIdentifier = accountId;
+    void setId(@Nullable final String id) {
+        mHomeOid = id;
     }
 
+    @NonNull
     @Override
-    public IAccountIdentifier getAccountIdentifier() {
-        return mAccountIdentifier;
-    }
+    public String getId() {
+        String id;
+
+        if (null != mRawIdToken && null != getClaims()) {
+            id = (String) getClaims().get(MicrosoftIdToken.OBJECT_ID);
+        } else {
+            id = mHomeOid;
+        }
 
-    /**
-     * Sets the home account id.
-     *
-     * @param homeAccountId The IAccountIdentifier to set.
-     */
-    void setHomeAccountIdentifier(final IAccountIdentifier homeAccountId) {
-        mHomeAccountIdentifier = homeAccountId;
+        return id;
     }
 
-    @Override
-    public IAccountIdentifier getHomeAccountIdentifier() {
-        return mHomeAccountIdentifier;
+    void setTenantId(@Nullable final String tenantId) {
+        mHomeTenantId = tenantId;
     }
 
-    /**
-     * Sets the username.
-     *
-     * @param username The username to set.
-     */
-    void setUsername(final String username) {
-        mUsername = username;
+    @Nullable
+    String getTenantId() {
+        return mHomeTenantId;
     }
 
-    @Override
-    public String getUsername() {
-        return mUsername;
+    @Nullable
+    String getHomeAccountId() {
+        return getId() + "." + mHomeTenantId;
     }
 
-    /**
-     * Sets the environment.
-     *
-     * @param environment The environment to set.
-     */
-    void setEnvironment(final String environment) {
+    void setEnvironment(@NonNull final String environment) {
         mEnvironment = environment;
     }
 
-    @Override
-    public String getEnvironment() {
+    @Nullable
+    String getEnvironment() {
         return mEnvironment;
     }
+
+    @Nullable
+    @Override
+    public Map<String, ?> getClaims() {
+        Map<String, ?> claims = null;
+
+        if (null != mRawIdToken) {
+            claims = getClaims(mRawIdToken);
```

Comment:
getClaims always returns a map... even if the mRawIdToken is null.... I noticed that rawIdToken was nullable... i'm assuming this is to address the scenario when no home account is available.

Reply:
That's right -- are you looking for a change here?

Reply:
I'll look at this. Will definitely add some supporting comments explaining what's going on

Reply:
I did some refactoring around this to better share code with `IDToken` in `common`.

Commits which address this can be found here:
- `common` -> [adf6d5bbe291da2df4b7c8fc5ff7157eb194dc13](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/461/commits/adf6d5bbe291da2df4b7c8fc5ff7157eb194dc13)
- `msal` -> ac88ab8a8fa651723d8fd5b884489d664c50c812

Reply:
Also improved the javadoc and nullability annots here: ac0b5714d96fe11bd7aa490621f75acf4a6b03d7